README
netmate v0.2.1 README
*********************

Netmate is a GTK+ program that shows network protocols headers in 32-bit
aligned fields, identical as they are represented in Request for Comments
(RFCs). It can be used to learn and teach about network protocols and to
understand their functionality and collaboration.

This tool supports input files in pcap(ng) format as they are created by
tcpdump, wireshark or other code based on libpcap.

Netmate is written in C and released under the GNU General Public License.
It was inspired by netdude, that is not further developed any more.

Currently the following network protocols are supported:
- Layer 2: Ethernet, SLL (Linux Cooked)
- Layer 3: ARP, ICMP, ICMPv6, IPv4, IPv6
- Layer 4: TCP, UDP

This tool is still under development! Please send any further wishes, feature
requests or problems in compiling and execution to ruport@f00l.de. Also send
me any more network protocols you want to be supported in future releases.

For further information visit the netmate homepage at http://f00l.de/netmate/.

USAGE
=====

./netmate filename

COPYRIGHT
=========

Copyright 2013-2016 Robert Krause

Netmate is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or any later version.

Netmate is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
Netmate. If not, see http://www.gnu.org/licenses/.


Changelog
2015-10-31 Robert Krause <Ruport@web.de>

 * added BSD local loopback as DLT type
 * removed separate ICMP data fields with one single field
 * updated code to match GTK 3.10 requirements

2013-11-17 Robert Krause <Ruport@web.de>

 * added full support for neighbor discovery protocol (NDP)
 * fixed GPLv3 license statement and added to all files
 * minor bug fixes

2013-11-10 Robert Krause <Ruport@web.de>

 * added resolution of header fields for all supported protocols
 * minor bug fixes

2013-10-28 Robert Krause <Ruport@web.de>

 * added tooltips for all supported protocols
 * improved Makefile
 * minor bug fixes

2013-10-19 Robert Krause <Ruport@web.de>

 * improved Makefile
 * added shortcut and icon
 * added README and man-page
 * compiles under windows
 * minor bug fixes

2013-10-05 Robert Krause <Ruport@web.de>

 * initial changelog
 * added columns source port, destination port, flags
 * columns are resizeable now
 * removed external GUI (netmate.ui)
 * minor fixes


win32.h
/*******************************************************************************
 *
 * Copyright (c) 2013-2016 Robert Krause (ruport@f00l.de)
 *
 * This file is part of Netmate.
 *
 * Netmate is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or any later version.
 *
 * Netmate is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Netmate. If not, see http://www.gnu.org/licenses/.
 *
 ******************************************************************************/

#define ETH_ALEN	6		/* Octets in one ethernet addr	 */

#define	ETHERTYPE_IP		0x0800		/* IP */
#define	ETHERTYPE_ARP		0x0806		/* Address resolution */
#define	ETHERTYPE_IPV6		0x86dd		/* IP protocol version 6 */

struct arphdr
  {
    unsigned short int ar_hrd;		/* Format of hardware address.  */
    unsigned short int ar_pro;		/* Format of protocol address.  */
    unsigned char ar_hln;		/* Length of hardware address.  */
    unsigned char ar_pln;		/* Length of protocol address.  */
    unsigned short int ar_op;		/* ARP opcode (command).  */
#if 0
    /* Ethernet looks like this : This bit is variable sized
       however...  */
    unsigned char __ar_sha[ETH_ALEN];	/* Sender hardware address.  */
    unsigned char __ar_sip[4];		/* Sender IP address.  */
    unsigned char __ar_tha[ETH_ALEN];	/* Target hardware address.  */
    unsigned char __ar_tip[4];		/* Target IP address.  */
#endif
  };

struct ether_header
{
  u_int8_t  ether_dhost[ETH_ALEN];	/* destination eth addr	*/
  u_int8_t  ether_shost[ETH_ALEN];	/* source ether addr	*/
  u_int16_t ether_type;		        /* packet type ID field	*/
} __attribute__ ((__packed__));

#define	IP_RF 0x8000			/* reserved fragment flag */
#define	IP_DF 0x4000			/* dont fragment flag */
#define	IP_MF 0x2000			/* more fragments flag */
#define	IP_OFFMASK 0x1fff		/* mask for fragmenting bits */

#define	IPOPT_COPY		0x80
#define	IPOPT_CLASS_MASK	0x60
#define	IPOPT_NUMBER_MASK	0x1f

struct iphdr
  {
#if __BYTE_ORDER == __LITTLE_ENDIAN
    u_int8_t ihl:4;
    u_int8_t version:4;
#elif __BYTE_ORDER == __BIG_ENDIAN
    u_int8_t ihl:4;
    u_int8_t version:4;
#else
# error	"Please fix <bits/endian.h>"
#endif
    u_int8_t tos;
    u_int16_t tot_len;
    u_int16_t id;
    u_int16_t frag_off;
    u_int8_t ttl;
    u_int8_t protocol;
    u_int16_t check;
    u_int32_t saddr;
    u_int32_t daddr;
    /*The options start here. */
  };

/* IPv6 address */
struct in6_addr
  {
    union
      {
	uint8_t	__u6_addr8[16];
	uint16_t __u6_addr16[8];
	uint32_t __u6_addr32[4];
      } __in6_u;
  };

struct ip6_hdr
  {
    union
      {
	struct ip6_hdrctl
	  {
	    uint32_t ip6_un1_flow;   /* 4 bits version, 8 bits TC,
					20 bits flow-ID */
	    uint16_t ip6_un1_plen;   /* payload length */
	    uint8_t  ip6_un1_nxt;    /* next header */
	    uint8_t  ip6_un1_hlim;   /* hop limit */
	  } ip6_un1;
	uint8_t ip6_un2_vfc;       /* 4 bits version, top 4 bits tclass */
      } ip6_ctlun;
    struct in6_addr ip6_src;      /* source address */
    struct in6_addr ip6_dst;      /* destination address */
  };

struct icmphdr
{
  u_int8_t type;		/* message type */
  u_int8_t code;		/* type sub-code */
  u_int16_t checksum;
  union
  {
    struct
    {
      u_int16_t	id;
      u_int16_t	sequence;
    } echo;			/* echo datagram */
    u_int32_t	gateway;	/* gateway address */
    struct
    {
      u_int16_t	__unused;
      u_int16_t	mtu;
    } frag;			/* path mtu discovery */
  } un;
};

struct icmp6_hdr
  {
    uint8_t     icmp6_type;   /* type field */
    uint8_t     icmp6_code;   /* code field */
    uint16_t    icmp6_cksum;  /* checksum field */
    union
      {
	uint32_t  icmp6_un_data32[1]; /* type-specific field */
	uint16_t  icmp6_un_data16[2]; /* type-specific field */
	uint8_t   icmp6_un_data8[4];  /* type-specific field */
      } icmp6_dataun;
  };

struct tcphdr
  {
    u_int16_t source;
    u_int16_t dest;
    u_int32_t seq;
    u_int32_t ack_seq;
#  if __BYTE_ORDER == __LITTLE_ENDIAN
    u_int16_t res1:4;
    u_int16_t doff:4;
    u_int16_t fin:1;
    u_int16_t syn:1;
    u_int16_t rst:1;
    u_int16_t psh:1;
    u_int16_t ack:1;
    u_int16_t urg:1;
    u_int16_t res2:2;
#  elif __BYTE_ORDER == __BIG_ENDIAN
    u_int16_t doff:4;
    u_int16_t res1:4;
    u_int16_t res2:2;
    u_int16_t urg:1;
    u_int16_t ack:1;
    u_int16_t psh:1;
    u_int16_t rst:1;
    u_int16_t syn:1;
    u_int16_t fin:1;
#  else
#   error "Adjust your <bits/endian.h> defines"
#  endif
    u_int16_t window;
    u_int16_t check;
    u_int16_t urg_ptr;
};

struct udphdr
{
  u_int16_t source;
  u_int16_t dest;
  u_int16_t len;
  u_int16_t check;
};


layer4.h
/*******************************************************************************
 *
 * Copyright (c) 2013-2016 Robert Krause (ruport@f00l.de)
 *
 * This file is part of Netmate.
 *
 * Netmate is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or any later version.
 *
 * Netmate is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Netmate. If not, see http://www.gnu.org/licenses/.
 *
 ******************************************************************************/

/******************************************************************************/
/* netmate layer4 protocols */
/******************************************************************************/

char *tcp_optkind(unsigned char id);
GtkGrid *tcp_grid(struct tcphdr *tcp, u_char *options);
GtkGrid *udp_grid(struct udphdr *udp);

/******************************************************************************/

char *tcp_optkind(unsigned char id) {
  switch (id) {
    case 0:
      return("End of Options");
    case 1:
      return("No Operation");
    case 2:
      return("Maximum segment size");
    case 3:
      return("Window scale");
    case 4:
      return("SACK permitted");
    case 5:
      return("Selective ACKnowledgement");
    case 8:
      return("Timestamp");
    case 14:
      return("TCP Alternate Checksum Request");
    case 15:
      return("TCP Alternate Checksum Data");
  }
  return("UNKNOWN");
}

GtkGrid *tcp_grid(struct tcphdr *tcp, u_char *options) {
  GtkGrid *grid;	/* the grid itself */
  char *label;		/* label of buttons to set */
  char *optdata;	/* option data */
  int x,y;			/* position pointer to next empty grid cell */
  int left;			/* bytes left for ipv4 options */
  int optlen;		/* length of options field */
  int optkind;		/* option kind */
  int i;			/* loop counter for raw data representation */

  /* init new empty grid */
  grid = GTK_GRID(gtk_grid_new());

  /* set columns to be uniform sized (for better bit size representation) */
  gtk_grid_set_column_homogeneous(grid, TRUE);

  /* allocate memory for button label */
  label = malloc(100);

  /* build bit indication topic line (0 1 2 .. 31) */
  for (x=0; x<32; x++) {
    sprintf(label, "%u", x);
    gtk_grid_attach(grid, gtk_label_new(label), x, 0, 1, 1);
  }

  /* set cell pointer to next empty grid cell */
  x=0;
  y=1;

  /* source port */
  sprintf(label, "Source Port: %u", htons(tcp->source));
  append_field(grid, &x, &y, sizeof(tcp->source)*8, label, TCP_SPORT);

  /* destination port */
  sprintf(label, "Destination Port: %u", htons(tcp->dest));
  append_field(grid, &x, &y, sizeof(tcp->dest)*8, label, TCP_DPORT);

  /* sequence number */
  sprintf(label, "Sequence Number: %u", htonl(tcp->seq));
  append_field(grid, &x, &y, sizeof(tcp->seq)*8, label, TCP_SEQ_NUM);

  /* acknowledgement number */
  sprintf(label, "Acknowledgement Number: %u", htonl(tcp->ack_seq));
  append_field(grid, &x, &y, sizeof(tcp->ack_seq)*8, label, TCP_ACK_NUM);

  /* data offset */
  sprintf(label, "Data Offset: %u (%u bytes)", tcp->doff, tcp->doff*4);
  append_field(grid, &x, &y, 4, label, TCP_DOFF);

  /* reserved (000) */
  if (tcp->res1 & 0x08) {
    append_field(grid, &x, &y, 1, "R", TCP_FLAG_RES);
  } else {
    append_field(grid, &x, &y, 1, "r", TCP_FLAG_RES);
  }
  if (tcp->res1 & 0x04) {
    append_field(grid, &x, &y, 1, "R", TCP_FLAG_RES);
  } else {
    append_field(grid, &x, &y, 1, "r", TCP_FLAG_RES);
  }
  if (tcp->res1 & 0x02) {
    append_field(grid, &x, &y, 1, "R", TCP_FLAG_RES);
  } else {
    append_field(grid, &x, &y, 1, "r", TCP_FLAG_RES);
  }

  /* NS */
  if (tcp->res1 & 0x01) {
    append_field(grid, &x, &y, 1, "NS", TCP_FLAG_NS);
  } else {
    append_field(grid, &x, &y, 1, "ns", TCP_FLAG_NS);
  }

  /* CWR */
  if (tcp->res2 & 0x02) {
    append_field(grid, &x, &y, 1, "CWR", TCP_FLAG_CWR);
  } else {
    append_field(grid, &x, &y, 1, "cwr", TCP_FLAG_CWR);
  }

  /* ECE */
  if (tcp->res2 & 0x01) {
    append_field(grid, &x, &y, 1, "ECE", TCP_FLAG_ECE);
  } else {
    append_field(grid, &x, &y, 1, "ece", TCP_FLAG_ECE);
  }

  /* URG */
  if (tcp->urg) {
    append_field(grid, &x, &y, 1, "URG", TCP_FLAG_URG);
  } else {
    append_field(grid, &x, &y, 1, "urg", TCP_FLAG_URG);
  }

  /* ACK */
  if (tcp->ack) {
    append_field(grid, &x, &y, 1, "ACK", TCP_FLAG_ACK);
  } else {
    append_field(grid, &x, &y, 1, "ack", TCP_FLAG_ACK);
  }

  /* PSH */
  if (tcp->psh) {
    append_field(grid, &x, &y, 1, "PSH", TCP_FLAG_PSH);
  } else {
    append_field(grid, &x, &y, 1, "psh", TCP_FLAG_PSH);
  }

  /* RST */
  if (tcp->rst) {
    append_field(grid, &x, &y, 1, "RST", TCP_FLAG_RST);
  } else {
    append_field(grid, &x, &y, 1, "rst", TCP_FLAG_RST);
  }

  /* SYN */
  if (tcp->syn) {
    append_field(grid, &x, &y, 1, "SYN", TCP_FLAG_SYN);
  } else {
    append_field(grid, &x, &y, 1, "syn", TCP_FLAG_SYN);
  }

  /* FIN */
  if (tcp->fin) {
    append_field(grid, &x, &y, 1, "FIN", TCP_FLAG_FIN);
  } else {
    append_field(grid, &x, &y, 1, "fin", TCP_FLAG_FIN);
  }

  /* window size */
  sprintf(label, "Window Size: %u", htons(tcp->window));
  append_field(grid, &x, &y, sizeof(tcp->window)*8, label, TCP_WINDOW_SIZE);

  /* checksum */
  sprintf(label, "Checksum: 0x%04x", htons(tcp->check));
  append_field(grid, &x, &y, sizeof(tcp->check)*8, label, TCP_CHECKSUM);

  /* urgent pointer */
  sprintf(label, "Urgent Pointer: %u", htons(tcp->urg_ptr));
  append_field(grid, &x, &y, sizeof(tcp->urg_ptr)*8, label, TCP_URGENT_POINTER);

  /* count bytes of option fields */
  left = (tcp->doff-0x05)*4;

  /* progress bytes until no option bytes left */
  while (left > 0) {
    /* get option type (first byte) */
    optkind = options[0];

    /* option kind */
    sprintf(label, "Option Kind: %u (%s)", optkind, tcp_optkind(optkind));
    append_field(grid, &x, &y, 8, label, TCP_OPTION_KIND);

    /* option dependent output (some options dont have a length field) */
    if ((optkind == 0x00) || (optkind == 0x01)) { /* EOO or NOP */
      /* option length is only 1 byte */
      optlen = 1;
    } else {
      /* options with variable length */

      /* option length (INCLUDING kind and length field) */
      optlen = options[1];
      sprintf(label, "Option Length: %u", optlen);
      append_field(grid, &x, &y, 8, label, TCP_OPTION_LENGTH);

      /* option has additional option data? */
      if (optlen > 2) {

        /* allocate memory for option data (*2 because of hex representation) */
        optdata = malloc(optlen*2);

        /* print bytes in hex format into array */
        for (i=0; i<optlen-2; ++i) sprintf(&optdata[i*2], "%02x", (unsigned int)options[i+2]);
        optdata[(optlen-2)*2] = 0x00;

        /* option data field */
        sprintf(label, "Option Data: 0x%s", optdata);
        append_field(grid, &x, &y, (optlen-2)*8, label, TCP_OPTION_DATA);

        /* free data */
        free(optdata);
      }
    }

    /* reduce length of field */
    left -= optlen;

    /* increase pointer to options header */
    options = options + optlen;
  }

  /* free memory of label */
  free(label);

  /* show ethernet grid (tab) */
  gtk_widget_show_all(GTK_WIDGET(grid));

  /* pass grid back to tab builder function */
  return(grid);
}

GtkGrid *udp_grid(struct udphdr *udp) {
  GtkGrid *grid;	/* the grid itself */
  char *label;		/* label of buttons to set */
  int x,y;			/* position pointer to next empty grid cell */

  /* init new empty grid */
  grid = GTK_GRID(gtk_grid_new());

  /* set columns to be uniform sized (for better bit size representation) */
  gtk_grid_set_column_homogeneous(grid, TRUE);

  /* allocate memory for button label */
  label = malloc(100);

  /* build bit indication topic line (0 1 2 .. 31) */
  for (x=0; x<32; x++) {
    sprintf(label, "%u", x);
    gtk_grid_attach(grid, gtk_label_new(label), x, 0, 1, 1);
  }

  /* set cell pointer to next empty grid cell */
  x=0;
  y=1;

  /* source port */
  sprintf(label, "Source Port: %u", htons(udp->source));
  append_field(grid, &x, &y, sizeof(udp->source)*8, label, UDP_SPORT);

  /* destination port */
  sprintf(label, "Destination Port: %u", htons(udp->dest));
  append_field(grid, &x, &y, sizeof(udp->dest)*8, label, UDP_DPORT);

  /* length */
  sprintf(label, "Length: %u", htons(udp->len));
  append_field(grid, &x, &y, sizeof(udp->len)*8, label, UDP_LENGTH);

  /* checksum */
  sprintf(label, "Checksum: 0x%02x", htons(udp->check));
  append_field(grid, &x, &y, sizeof(udp->check)*8, label, UDP_CHECKSUM);

  /* free memory of label */
  free(label);

  /* show ethernet grid (tab) */
  gtk_widget_show_all(GTK_WIDGET(grid));

  /* pass grid back to tab builder function */
  return(grid);
}




layer2.h
/*******************************************************************************
 *
 * Copyright (c) 2013-2016 Robert Krause (ruport@f00l.de)
 *
 * This file is part of Netmate.
 *
 * Netmate is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or any later version.
 *
 * Netmate is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Netmate. If not, see http://www.gnu.org/licenses/.
 *
 ******************************************************************************/

/******************************************************************************/
/* netmate layer2 protocols */
/******************************************************************************/

char *hardwaretype(unsigned short id);
char *ethertype(unsigned short id);
GtkGrid *ethernet_grid(struct ether_header *eth);	/* ethernet */
GtkGrid *sll_grid(struct sll_header *sll);		/* ssl (linux cooked) */

/******************************************************************************/

#define NULL_IPV4 2
#define NULL_IPV6_24 24
#define NULL_IPV6_28 28
#define NULL_IPV6_30 30

/*
   BSD loopback encapsulation; the link layer header is a 4-byte field, in
   host byte order, containing a value of 2 for IPv4 packets, a value of
   either 24, 28, or 30 for IPv6 packets, a value of 7 for OSI packets, or a
   value of 23 for IPX packets. All of the IPv6 values correspond to IPv6
   packets; code reading files should check for all of them.
*/
struct null_header {
  u_int32_t link_layer;      /* link layer */
};

/* taken from http://www.tcpdump.org/linktypes.html */
char *nulltype(unsigned int id) {
  switch (id) {
    case 2:
      return("IPV4");
    case 7:
      return("OSI");
    case 23:
      return("IPX");
    case 24:
    case 28:
    case 30:
      return("IPV6");
  }
  return("UNKNOWN");
}


/* taken from http://www.tcpdump.org/linktypes.html */
char *hardwaretype(unsigned short id) {
  switch (id) {
    case 0:
      return("NULL");
    case 1:
      return("ETHERNET");
    case 3:
      return("AX25");
    case 6:
      return("IEEE802_5");
    case 7:
      return("ARCNET_BSD");
    case 8:
      return("SLIP");
    case 9:
      return("PPP");
    case 10:
      return("FDDI");
    case 50:
      return("PPP_HDLC");
    case 51:
      return("PPP_ETHER");
    case 100:
      return("ATM_RFC1483");
    case 101:
      return("RAW");
    case 104:
      return("C_HDLC");
    case 105:
      return("IEEE802_11");
    case 107:
      return("FRELAY");
    case 108:
      return("LOOP");
    case 113:
      return("LINUX_SLL");
    case 114:
      return("LTALK");
    case 117:
      return("PFLOG");
    case 119:
      return("IEEE802_11_PRISM");
    case 122:
      return("IP_OVER_FC");
    case 123:
      return("SUNATM");
    case 127:
      return("IEEE802_11_RADIOTAP");
    case 129:
      return("ARCNET_LINUX");
    case 138:
      return("APPLE_IP_OVER_IEEE1394");
    case 139:
      return("MTP2_WITH_PHDR");
    case 140:
      return("MTP2");
    case 141:
      return("MTP3");
    case 142:
      return("SCCP");
    case 143:
      return("DOCSIS");
    case 144:
      return("LINUX_IRDA");
    case 163:
      return("IEEE802_11_AVS");
    case 165:
      return("BACNET_MS_TP");
    case 166:
      return("PPP_PPPD");
    case 169:
      return("GPRS_LLC");
    case 177:
      return("LINUX_LAPD");
    case 187:
      return("BLUETOOTH_HCI_H4");
    case 189:
      return("USB_LINUX");
    case 192:
      return("PPI");
    case 195:
      return("IEEE802_15_4");
    case 196:
      return("SITA");
    case 197:
      return("ERF");
    case 201:
      return("BLUETOOTH_HCI_H4_WITH_PHDR");
    case 202:
      return("AX25_KISS");
    case 203:
      return("LAPD");
    case 204:
      return("PPP_WITH_DIR");
    case 205:
      return("C_HDLC_WITH_DIR");
    case 206:
      return("FRELAY_WITH_DIR");
    case 209:
      return("IPMB_LINUX");
    case 215:
      return("IEEE802_15_4_NONASK_PHY");
    case 220:
      return("USB_LINUX_MMAPPED");
    case 224:
      return("FC_2");
    case 225:
      return("FC_2_WITH_FRAME_DELIMS");
    case 226:
      return("IPNET");
    case 227:
      return("CAN_SOCKETCAN");
    case 228:
      return("IPV4");
    case 229:
      return("IPV6");
    case 230:
      return("IEEE802_15_4_NOFCS");
    case 231:
      return("DBUS");
    case 235:
      return("DVB_CI");
    case 236:
      return("MUX27010");
    case 237:
      return("STANAG_5066_D_PDU");
    case 239:
      return("NFLOG");
    case 240:
      return("NETANALYZER");
    case 241:
      return("NETANALYZER_TRANSPARENT");
    case 242:
      return("IPOIB");
    case 243:
      return("MPEG_2_TS");
    case 244:
      return("NG40");
    case 245:
      return("NFC_LLCP");
    case 247:
      return("INFINIBAND");
    case 248:
      return("SCTP");
    case 249:
      return("USBPCAP");
    case 250:
      return("RTAC_SERIAL");
    case 251:
      return("BLUETOOTH_LE_LL");
  }
  return("UNKNOWN");
}

/* taken from http://en.wikipedia.org/wiki/EtherType */
char *ethertype(unsigned short id) {
  switch (id) {
    case 0x0800:
      return("IPv4");
    case 0x0806:
      return("ARP");
    case 0x0842:
      return("Wake-on-LAN");
    case 0x22F3:
      return("IETF TRILL Protocol");
    case 0x6003:
      return("DECnet Phase IV");
    case 0x8035:
      return("RARP");
    case 0x809B:
      return("AppleTalk (Ethertalk)");
    case 0x80F3:
      return("AARP");
    case 0x8100:
      return("VLAN-tagged frame (IEEE 802.1Q)");
    case 0x8137:
      return("IPX");
    case 0x8138:
      return("IPX");
    case 0x8204:
      return("QNX Qnet");
    case 0x86DD:
      return("IPv6");
    case 0x8808:
      return("Ethernet flow control");
    case 0x8809:
      return("Slow Protocols (IEEE 802.3)");
    case 0x8819:
      return("CobraNet");
    case 0x8847:
      return("MPLS unicast");
    case 0x8848:
      return("MPLS multicast");
    case 0x8863:
      return("PPPoE Discovery Stage");
    case 0x8864:
      return("PPPoE Session Stage");
    case 0x8870:
      return("Jumbo Frames");
    case 0x887B:
      return("HomePlug 1.0 MME");
    case 0x888E:
      return("EAP over LAN (IEEE 802.1X)");
    case 0x8892:
      return("PROFINET Protocol");
    case 0x889A:
      return("HyperSCSI");
    case 0x88A2:
      return("ATA over Ethernet");
    case 0x88A4:
      return("EtherCAT Protocol");
    case 0x88A8:
      return("Provider Bridging (IEEE 802.1ad)");
    case 0x88AB:
      return("Ethernet Powerlink");
    case 0x88CC:
      return("LLDP");
    case 0x88CD:
      return("SERCOS III");
    case 0x88E1:
      return("HomePlug AV MME");
    case 0x88E3:
      return("Media Redundancy Protocol (IEC62439-2)");
    case 0x88E5:
      return("MAC security (IEEE 802.1AE)");
    case 0x88F7:
      return("Precision Time Protocol (IEEE 1588)");
    case 0x8902:
      return("IEEE 802.1ag Connectivity Fault Management (CFM) Protocol");
    case 0x8906:
      return("Fibre Channel over Ethernet (FCoE)");
    case 0x8914:
      return("FCoE Initialization Protocol");
    case 0x8915:
      return("RDMA over Converged Ethernet (RoCE)");
    case 0x892F:
      return("High-availability Seamless Redundancy (HSR)");
    case 0x9000:
      return("Ethernet Configuration Testing Protocol");
    case 0x9100:
      return("Q-in-Q");
    case 0xCAFE:
      return("Veritas Low Latency Transport (LLT)");
  }
  return("UNKNOWN");
}

GtkGrid *sll_grid(struct sll_header *sll) {
  GtkGrid *grid;	/* the grid itself */
  int x, y;			/* position of next empty grid cell */
  char *label;		/* label of buttons to set */

  /* init a new empty grid */
  grid = GTK_GRID(gtk_grid_new());

  /* set columns to be uniform sized (for better bit size representation) */
  gtk_grid_set_column_homogeneous(grid, TRUE);

  /* allocate memory for button label */
  label = malloc(100);

  /* build bit indication topic line (0 1 2 .. 31) */
  for (x=0; x<32; x++) {
    sprintf(label, "%u", x);
    gtk_grid_attach(grid, gtk_label_new(label), x, 0, 1, 1);
  }

  /* set cell pointer to next empty grid cell */
  x=0;
  y=1;

  /* Packet Type */
  switch (htons(sll->sll_pkttype)) {
    case 0:
      sprintf(label, "Packet Type: %u (%s)", htons(sll->sll_pkttype), "Sent to us (unicast)");
      break;
    case 1:
      sprintf(label, "Packet Type: %u (%s)", htons(sll->sll_pkttype), "Sent to us (broadcast)");
      break;
    case 2:
      sprintf(label, "Packet Type: %u (%s)", htons(sll->sll_pkttype), "Sent to us (multicast)");
      break;
    case 3:
      sprintf(label, "Packet Type: %u (%s)", htons(sll->sll_pkttype), "Sent to someone else (unicast)");
      break;
    case 4:
      sprintf(label, "Packet Type: %u (%s)", htons(sll->sll_pkttype), "Sent by us");
      break;
    default:
      sprintf(label, "Packet Type: %u (%s)", htons(sll->sll_pkttype), "UNKNOWN");
  }
  append_field(grid, &x, &y, sizeof(sll->sll_pkttype)*8, label, SLL_PACKET_TYPE);

  /* ARP Header Type */
  sprintf(label, "ARPHRD_ Type: %u (%s)", htons(sll->sll_hatype), hardwaretype(htons(sll->sll_hatype)));
  append_field(grid, &x, &y, sizeof(sll->sll_hatype)*8, label, SLL_ARPHRD_TYPE);

  /* Link-layer Address Length */
  sprintf(label, "Link-layer Address Length: %u", htons(sll->sll_halen));
  append_field(grid, &x, &y, sizeof(sll->sll_halen)*8, label, SLL_LLA_LENGTH);

  /* Link-layer Address */
  sprintf(label, "Link-layer Address: %02x:%02x:%02x:%02x:%02x:%02x:%02x:%02x", sll->sll_addr[0], sll->sll_addr[1], sll->sll_addr[2], sll->sll_addr[3], sll->sll_addr[4], sll->sll_addr[5], sll->sll_addr[6], sll->sll_addr[7]);
  append_field(grid, &x, &y, sizeof(sll->sll_addr)*8, label, SLL_LLA);

  /* Upper Layer Protocol */
  sprintf(label, "Protocol Type: 0x%04x (%s)", htons(sll->sll_protocol), ethertype(htons(sll->sll_protocol)));
  append_field(grid, &x, &y, sizeof(sll->sll_protocol)*8, label, SLL_PROTOCOL);

  /* free memory of label */
  free(label);

  /* show ethernet grid (tab) */
  gtk_widget_show_all(GTK_WIDGET(grid));

  /* return grid to tab builder */
  return(grid);
}

GtkGrid *ethernet_grid(struct ether_header *eth) {
  GtkGrid *grid;	/* the grid itself */
  int x, y;			/* position pointer to next empty grid cell */
  char *label;		/* label of buttons to set */

  /* init new empty grid */
  grid = GTK_GRID(gtk_grid_new());

  /* set columns to be uniform sized (for better bit size representation) */
  gtk_grid_set_column_homogeneous(grid, TRUE);

  /* allocate memory for button label */
  label = malloc(100);

  /* build bit indication topic line (0 1 2 .. 31) */
  for (x=0; x<32; x++) {
    sprintf(label, "%u", x);
    gtk_grid_attach(grid, gtk_label_new(label), x, 0, 1, 1);
  }

  /* set cell pointer to next empty grid cell */
  x=0;
  y=1;

  /* destination mac */
  sprintf(label, "Destination: %02x:%02x:%02x:%02x:%02x:%02x", eth->ether_dhost[0], eth->ether_dhost[1], eth->ether_dhost[2], eth->ether_dhost[3], eth->ether_dhost[4], eth->ether_dhost[5]);
  append_field(grid, &x, &y, sizeof(eth->ether_dhost)*8, label, ETHERNET_DESTINATION);

  /* source mac */
  sprintf(label, "Source: %02x:%02x:%02x:%02x:%02x:%02x", eth->ether_shost[0], eth->ether_shost[1], eth->ether_shost[2], eth->ether_shost[3], eth->ether_shost[4], eth->ether_shost[5]);
  append_field(grid, &x, &y, sizeof(eth->ether_shost)*8, label, ETHERNET_DESTINATION);

  /* upper layer protocol */
  sprintf(label, "Type: 0x%04x (%s)", htons(eth->ether_type), ethertype(htons(eth->ether_type)));
  append_field(grid, &x, &y, sizeof(eth->ether_type)*8, label, ETHERNET_TYPE);

  /* free memory of label */
  free(label);

  /* show ethernet grid (tab) */
  gtk_widget_show_all(GTK_WIDGET(grid));

  /* return grid to tab builder */
  return(grid);
}

GtkGrid *null_grid(struct null_header *null) {
  GtkGrid *grid;	/* the grid itself */
  int x, y;		/* position pointer to next empty grid cell */
  char *label;		/* label of buttons to set */

  /* init new empty grid */
  grid = GTK_GRID(gtk_grid_new());

  /* set columns to be uniform sized (for better bit size representation) */
  gtk_grid_set_column_homogeneous(grid, TRUE);

  /* allocate memory for button label */
  label = malloc(100);

  /* build bit indication topic line (0 1 2 .. 31) */
  for (x=0; x<32; x++) {
    sprintf(label, "%u", x);
    gtk_grid_attach(grid, gtk_label_new(label), x, 0, 1, 1);
  }

  /* set cell pointer to next empty grid cell */
  x=0;
  y=1;

  /* link layer header */
  sprintf(label, "Link layer: %u (%s)", null->link_layer, nulltype(null->link_layer));
  append_field(grid, &x, &y, sizeof(null->link_layer)*8, label, NULL_LINK_LAYER);

  /* free memory of label */
  free(label);

  /* show ethernet grid (tab) */
  gtk_widget_show_all(GTK_WIDGET(grid));

  /* return grid to tab builder */
  return(grid);
}
